oauth2client/flask.py

```diff
@@ -0,0 +1,299 @@
+        bp = Blueprint('oauth2', __name__)
+        bp.add_url_rule('/oauth2authorize', 'authorize', self.authorize_view)
+        bp.add_url_rule('/oauth2logout', 'logout', self.logout_view)
+        bp.add_url_rule('/oauth2callback', 'callback', self.callback_view)
+
+        return bp
+
+    def authorize_view(self):
+        return_url = request.args.get('return_url')
+        if not return_url:
+            return_url = request.referrer or '/'
+        session['oauth2return'] = return_url
+
+        flow = self._make_flow()
+        auth_url = flow.step1_get_authorize_url()
+        return redirect(auth_url)
+
+    def logout_view(self):
+        session.pop('oauth2credentials', None)
+        session.pop('oauth2profile', None)
+        session.pop('oauth2userid', None)
+
+        if self.logout_callback:
+            self.logout_callback()
+
+        return redirect(request.referrer or '/')
+
+    def callback_view(self):
+        # Check the CSRF token.
+        if request.args.get('state', '') != session.pop('oauth2state', None):
+            return 'Invalid request state', 400
+
+        flow = self._make_flow()
+
+        # Exchange the auth code for credentials.
+        try:
+            credentials = flow.step2_exchange(request.args['code'])
+        except FlowExchangeError as e:
+            current_app.logger.exception(e)
+            return e.message, 400
+
+        # Save the credentials to the session.
+        session['oauth2credentials'] = credentials.to_json()
+
+        # Get the user's profile from Google and save it to the session.
+        if 'profile' in self.scopes:
+            session['oauth2profile'] = self._request_user_info()
+
+        if self.authorize_callback:
+            self.authorize_callback()
+
+        return redirect(session.pop('oauth2return', '/'))
+
+    @property
+    def credentials(self):
+        ctx = _app_ctx_stack.top
+
+        if not hasattr(ctx, 'oauth2_credentials'):
+            serialized = session.get('oauth2credentials')
+
+            if not serialized:
+                return None
+
+            ctx.oauth2_credentials = Credentials.new_from_json(serialized)
+
+        return ctx.oauth2_credentials
+
+    def has_credentials(self):
+        return self.credentials and not self.credentials.invalid
+
+    @property
+    def email(self):
+        """Returns the user's email address.
+
+        The email address is provided by the current credentials' id_token. The
+        use can change their email, it's more appropriate to use user_id for
+        a unique, user-specific identifier.
+        """
+        if not self.credentials:
+            return None
+        return self.credentials.id_token['email']
+
+    @property
+    def user_id(self):
+        """Returns the user's email address.
+
+        The id is provided by the current credentials' id_token.
+        """
+        if not self.credentials:
+            return None
+        return self.credentials.id_token['sub']
```

Comment:
I don't think it's possible to get an id_token without this field.

Caching here is overkill, credentials are already cached in flask context.